feat: add `toSorted` method to `array/fixed-endian-factory`

ahmad-kashkoush opened the pull request:
Resolves #3160 

## Description

> What is the purpose of this pull request?

This pull request:

-   adds support for a `toSorted` method to [@stdlib/array/fixed-endian-factory](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/fixed-endian-factory).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3160 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ahmad-kashkoush committed:
feat: add logic, tests, and benchmarks to toSorted method

ahmad-kashkoush committed:
docs: add documentation for toSorted method

ahmad-kashkoush committed:
Merge branch 'develop' of github.com:stdlib-js/stdlib into endian-toSorted

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_sorted.js:
```diff
@@ -0,0 +1,60 @@
+		}
+	}
+	b.toc();
+	if ( !isTypedArray( out ) ) {
```
This is not a valid check, as `out` is not an actual typed array instance.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_sorted.length.js:
```diff
@@ -0,0 +1,118 @@
+			}
+		}
+		b.toc();
+		if ( !isTypedArray( out ) ) {
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -669,6 +669,70 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+			throw new TypeError( format( 'invalid invocation. `this` is not %s %s.', CHAR2ARTICLE[ dtype[0] ], CTOR_NAME ) );
+		}
+		if ( arguments.length === 0 ) {
+			compareFcn = function defaultCompareFcn( a, b ) {
```
We don't typically use function expressions, only declarations. I suggest refactoring accordingly. It is also not necessary to have this function in a nested scope and can be lifted to the top most scope.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -669,6 +669,70 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+			minVal = obuf[ GETTER ]( i * BYTES_PER_ELEMENT, this._isLE );
+			for ( j = i+1; j < len; j++ ) {
+				val = obuf[ GETTER ]( j * BYTES_PER_ELEMENT, this._isLE );
+				if ( compareFcn( minVal, val ) > 0 ) {
```
Rather than rolling your own sort function, copy to a temporary generic array, sort using the built-in `#.sort` method, and then copy to an output array.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/test/test.to_sorted.js:
```diff
@@ -0,0 +1,211 @@
+	expected = [ 1.0, 2.0, 3.0, 4.0, 5.0 ];
+	out = arr.toSorted( compareFcn );
+	values = [];
+	fcn();
```
Rather than use `fcn` to generate a values array, just use `array/base/assert/has-same-values` and compare to an expected `Float64ArrayFE`.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/test/test.to_sorted.js:
```diff
@@ -0,0 +1,211 @@
+	expected = [ 1.0, 2.0, 3.0, 4.0, 5.0 ];
+	out = arr.toSorted();
+	values = [];
+	fcn();
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/README.md:
```diff
@@ -607,6 +607,39 @@ var idx = arr.lastIndexOf( 5.0 );
+var arr = new Float64ArrayFE( 'little-endian', [ 3.0, 1.0, 2.0 ] );
+// returns <Float64ArrayFE>
+
+var out=arr.toSorted( compareFcn );
```
```suggestion
var out = arr.toSorted( compareFcn );
```

You need spaces around operators here and below.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/README.md:
```diff
@@ -607,6 +607,39 @@ var idx = arr.lastIndexOf( 5.0 );
+var out=arr.toSorted( compareFcn );
+// returns <Float64ArrayFE>
+
+var v=out.get(0);
```
```suggestion
var v = out.get( 0 );
```

Spaces around arguments here and below.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/README.md:
```diff
@@ -607,6 +607,39 @@ var idx = arr.lastIndexOf( 5.0 );
 
+<a name="method-toSorted"></a>
+
+#### TypedArrayFE.prototype.toSorted( compareFcn )
```
The `compareFcn` is optional.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/README.md:
````diff
@@ -607,6 +607,39 @@ var idx = arr.lastIndexOf( 5.0 );
 // returns -1
 ```
 
+<a name="method-toSorted"></a>
````
This method should be inserted into the docs such that methods are listed in alphabetical order.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -669,6 +669,70 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+	* @throws {TypeError} first argument must be a function
+	* @returns {TypedArray} sorted array
+	*/
+	setReadOnly( TypedArray.prototype, 'toSorted', function sort( compareFcn ) {
```
Insert methods in alphabetical order.

ahmad-kashkoush committed:
refactor: update toSorted method

- update method logic and insert it into correct place in both main.js and readme
- update type checking in benchmarks
- use hasSameValues in test file

ahmad-kashkoush committed:
Merge remote-tracking branch 'rem/develop' into endian-toSorted

ahmad-kashkoush commented:
@kgryte @Planeshifter  I have applyed all changes you've suggested in this PR and also in #3298

ahmad-kashkoush commented:
@kgryte may you give it a review?
